src/compiler/parser.ts

```diff
@@ -2542,6 +2537,26 @@ namespace ts {
             return false;
         }
 
+        function parseTypeOrTypePredicate(): TypeNode {
+            const typePredicateVariable = tryParse(() => {

```

Comment:
Only call `tryParse` if a call to `isIdentifier()` is true. No point in incurring the `tryParse` overhead if it isn't. Also, don't pass an arrow function to `tryParse`. Instead, make a function call `nextTokenIsIsKeyword` or some such and pass that. Avoids creating a function object on every invocation.